Added a of Gen AI landing page (#20253)
openvinotoolkit/openvino: docs/articles_en/openvino_workflow/gen_ai.md

# Optimize and Deploy Generative AI Models {#gen_ai_guide}

@sphinxdirective

OpenVINO toolkit could be used to accelerate Generative AI use cases as those mostly rely on model inference. When it comes to Generative Models, OpenVINO supports:

* Conversion, optimization and inference for Text, Image and Audio generative models (for example, llama2, MPT, OPT, Stable Diffusion, Stable Diffusion XL, etc.) 
* Int8 weights compression for text generation models 
* Reduced storage format (fp16 precision for non-compressed models and int8 for compressed models) 
* Inference on CPU and GPU platforms, including Intel integrated and discrete ARC GPUs 
 

> There are two main paths how OpenVINO could be used to run Generative AI use cases:

Reviewer: I don't know if we want to complicate this document, and I don't know the status of that for LLMs, but for stable diffusion, toch.compile() is also a good option.

Author: It is experimental for LLMs still from what I know. But we can improve this in the next version when it is more mature.

We recommend starting with use of Hugging Face frameworks, initially experimenting with different models and scenarios, ding your fit and then convert to OpenVINO native APIs if there is a demand.  

Within Optimum-Intel product there are interfaces that allow to optimize model with use of `Neural Network Compression Framework (NNCF) <https://github.com/openvinotoolkit/nncf>`__ (compress weights) and export model to OpenVINO model format to be used in native API applications. 

Table below summarizes differences of both Hugging Face and Native approaches. 

.. list-table::
   :widths: 20 25 55
   :header-rows: 1

   * -  
     - Hugging Face through OpenVINO
     - OpenVINO Native API
   * - Model support
     - Wide set of Models
     - Wide set of Models
   * - APIs
     - Python (Hugging Face API)
     - Python, C++ (OpenVINO API)
   * - Model Format
     - Source Framework
     - OpenVINO
   * - Inference code
     - Hugging Face based
     - Custom inference pipelines
   * - Time to inference
     - Slow
     - Fast

>    * - Additional dependencies
>      - As much as Hugging Face has
>      - Numpy

Reviewer: ```suggestion
   * - Additional Python dependencies
```
openvino-telemetry is also a dependency of OpenVINO runtime Python API. And for C++, we require some libraries to be installed too. These libraries are included in the Python wheels, but when not using wheels they need to be installed.

Author: In fact, telemetry is a part of OpenVINO distribution that is just extracted into a separate package for convenience. I agree that it does not sound fully correct. Maybe we should rephrase it to "lightweight (e.g. numpy, boost,, etc.)". @yury-gorbachev, what do you think?

Reviewer 2: I agree- it should not be mentioned, this is part of OV distribution.

     - OpenVINO samples and notebooks
   * - Performance
     - Good
     - Best

Running Generative AI models using Hugging Face Optimium-Intel 
##############################################################

Prerequisits
+++++++++++++++++++++++++++

* Create a Python environment
* Install Optimium-Intel running:

.. code-block:: console

    pip install optimum[openvino,nncf]

To start using OpenVINO as a backend for Hugging Face you should change original Hugging Face code in two places. 

.. code-block:: diff

    -from transformers import AutoModelForCausalLM
    +from optimum.intel. import OVModelForCausalLM

    model_id = "meta-llama/Llama-2-7b-chat-hf"
    -model = AutoModelForCausalLM.from_pretrained(model_id)
    +model = OVModelForCausalLM.from_pretrained(model_id, export=True)

Alternativly, you can download and convert model using CLI interface: ``optimum-cli export openvino --model meta-llama/Llama-2-7b-chat-hf llama_openvino``
In this case, you can load the converted model in OpenVINO representation directly from the disk:

.. code-block:: python

    model_id = "llama_openvino"
    model = OVModelForCausalLM.from_pretrained(model_id)

> To select inference device use `.to()` method. The device naming convention is the same as in OpenVINO native API:
> 

Reviewer: ```suggestion
By default, inference will run on CPU. To select a different inference device, for example GPU, add `device="GPU"` to the `.from_pretrained()` call. To switch to a different device after the model has been loaded, use the `.to()` method. The device naming convention is the same as in OpenVINO native API:

```

# Optimize and Deploy Generative AI Models {#gen_ai_guide}

@sphinxdirective

Generative AI is an innovative technique that creates new data, such as text, images, video, or audio, using neural networks. OpenVINO accelerates Generative AI use cases as they mostly rely on model inference, allowing for faster development and better performance. When it comes to generative models, OpenVINO supports:

* Conversion, optimization and inference for text, image and audio generative models, for example, Llama 2, MPT, OPT, Stable Diffusion, Stable Diffusion XL, etc. 
* Int8 weight compression for text generation models. 
* Storage format reduction (fp16 precision for non-compressed models and int8 for compressed models). 
* Inference on CPU and GPU platforms, including Intel integrated and discrete ARC GPUs. 
 

OpenVINO offers two main paths for Generative AI use cases:

* Using OpenVINO as a backend for Hugging Face frameworks (transformers, diffusers) through the `Optimum Intel <https://huggingface.co/docs/optimum/intel/inference>`__ extension.
* Using OpenVINO native APIs (Python and C++) with custom pipeline code. 

 

> In both cases, OpenVINO runtime and tools are used, the difference is mostly in the preferred API and the final solution's footprint. Native APIs enable the use of generative models in C++ applications, ensure minimal runtime dependencies, and minimize application footprint. The Native APIs approach requires the implementation of glue code (generation loop, text tokenization, or scheduler functions), which is hidden as internal implementation details within Hugging Face libraries. 

Author: "...for a better developer experience." sounds more suitable to me.

It is recommended to start with Hugging Face frameworks. Experiment with different models and scenarios to find your fit, and then consider converting to OpenVINO native APIs based on your specific requirements. 

Within Optimum-Intel product there are interfaces that allow to optimize (compress weights) model with use of `Neural Network Compression Framework (NNCF) <https://github.com/openvinotoolkit/nncf>`__ and export model to OpenVINO model format to be used in native API applications. 

The table below summarizes differences of both Hugging Face and Native approaches. 

.. list-table::
   :widths: 20 25 55
   :header-rows: 1

   * -  
     - Hugging Face through OpenVINO
     - OpenVINO Native API
   * - Model support

>      - Wide set of Models
>      - Wide set of Models

Reviewer 3: ```suggestion
     - Broad set of Models
     - Broad set of Models
```

   * - APIs
     - Python (Hugging Face API)
     - Python, C++ (OpenVINO API)
   * - Model Format
     - Source Framework / OpenVINO
     - OpenVINO
   * - Inference code
     - Hugging Face based
     - Custom inference pipelines
   * - Additional dependencies

>      - As much as Hugging Face has

Reviewer: It is all Hugging Face dependencies. Technically it depends on the model too, for some models you need extra dependencies. But the idea here is to show that (for now at least) the HF path has many dependencies.

     - Ligthweight (e.g. numpy, boost, etc.)
   * - Application footprint
     - Large
     - Small
   * - Pre/post-processing and glue code
     - Available at Hugging Face OOB
     - OpenVINO samples and notebooks
   * - Performance
     - Good
     - Best

Running Generative AI models using Hugging Face Optimium-Intel 
##############################################################

Prerequisites
+++++++++++++++++++++++++++

* Create a Python environment
* Install Optimium-Intel running:

.. code-block:: console

    pip install optimum[openvino,nncf]

> To start using OpenVINO as a backend for Hugging Face you should change original Hugging Face code in two places. 

Reviewer 4: ```suggestion
To start using OpenVINO as a backend for Hugging Face, change the original Hugging Face code in two places:
```

.. code-block:: diff

    -from transformers import AutoModelForCausalLM
    +from optimum.intel. import OVModelForCausalLM

    model_id = "meta-llama/Llama-2-7b-chat-hf"
    -model = AutoModelForCausalLM.from_pretrained(model_id)
    +model = OVModelForCausalLM.from_pretrained(model_id, export=True)

> After that, you can call ``.save_pretrained()`` method to save model to the folder in the OpenVINO Intermediate Representation and use it further.

Reviewer 4: ```suggestion
After that, you can call the ``.save_pretrained()`` method to save the model to the folder in the OpenVINO Intermediate Representation and use it further.
```

.. code-block:: python

    model.save_pretrained(model_dir)

> Alternatively, you can download and convert model using CLI interface: ``optimum-cli export openvino --model meta-llama/Llama-2-7b-chat-hf llama_openvino``
> In this case, you can load the converted model in OpenVINO representation directly from the disk:

Reviewer 4: ```suggestion
Alternatively, you can download and convert the model using CLI interface: ``optimum-cli export openvino --model meta-llama/Llama-2-7b-chat-hf llama_openvino``.
In this case, you can load the converted model in OpenVINO representation directly from the disk:
```

.. code-block:: python

    model_id = "llama_openvino"
    model = OVModelForCausalLM.from_pretrained(model_id)

By default, inference will run on CPU. To select a different inference device, for example GPU, add `device="GPU"` to the `.from_pretrained()` call. To switch to a different device after the model has been loaded, use the `.to()` method. The device naming convention is the same as in OpenVINO native API:

.. code-block:: python

    model.to("GPU")

> Optimum-Intel API also provides out-of-the-box model optimization through weights compression using NNCF which substantially reduces the model footdpring and inference latency:

Reviewer 4: not sure but I checked that Optimum Intel is spelled as two words in other docs, correct me if that's wrong

> Optimum-Intel API also provides out-of-the-box model optimization through weights compression using NNCF which substantially reduces the model footdpring and inference latency:

Reviewer 4: ```suggestion
Optimum Intel API also provides out-of-the-box model optimization through weight compression using NNCF which substantially reduces the model footprint and inference latency:
```

Reviewer: I would keep "weights compression" since it's the name of the method and it should be easily findable with that name.

Reviewer 3: I see Helena's point, but that's not how it's typically referred to in the industry. As far as I can tell, the concept is "weight compression", and the action/verb is "compress weights". In terms of method/API naming I would expect something like `weight_compression()` or `compress_weights()`. If that is not the case today, and there's still time to change it, then I would recommend fixing the API name.

.. code-block:: python

>     model = OVModelForCausalLM.from_pretrained(model_id, export=True, load_in_8bit=True)

Author: OVQuantizer is still there and it is exposed in the HF docs. Here we who the most simplest way.
`load_in_8bit` is exactly the same interface that already exists in AutoModelForCausalLM. We just added the implementation for OpenVINO in Optimum-Intel.

>     model = OVModelForCausalLM.from_pretrained(model_id, export=True, load_in_8bit=True)

Reviewer 3: Makes sense then, if we're following HF practice. Still feels clumsy to me ;-)

> Wieght compression is applied by default to models larger than 1 billion parameters and also available for CLI interface as ``--int8`` option.

Reviewer 4: ```suggestion
Weight compression is applied by default to models larger than one billion parameters and is also available for CLI interface as the ``--int8`` option.
```

> Working with models tuned with LoRA

Reviewer 4: ```suggestion
Working with Models Tuned with LoRA
```

++++++++++++++++++++++++++++++++++++

> Low-ranking Adaptation (LoRA) is a popular method to tune Generative AI models to a downstream task or custom data. But it requires some extra steps to be done for efficient deployment using the Hugging Face API. Namely, the trained adapters should be fused into the baseline model to avoid extra computation. This is how it can be done for LLMs:

Reviewer 4: ```suggestion
Low-rank Adaptation (LoRA) is a popular method to tune Generative AI models to a downstream task or custom data. However, it requires some extra steps to be done for efficient deployment using the Hugging Face API. Namely, the trained adapters should be fused into the baseline model to avoid extra computation. This is how it can be done for Large Language Models (LLMs):
```

> Low-ranking Adaptation (LoRA) is a popular method to tune Generative AI models to a downstream task or custom data. But it requires some extra steps to be done for efficient deployment using the Hugging Face API. Namely, the trained adapters should be fused into the baseline model to avoid extra computation. This is how it can be done for LLMs:

Author: I don't think this is good to put too many links to 3rd party docs. It is just yet another way to leave OpenVINO docs. I think this part is useless for those who don't know what LoRA is about.
But let's see what our documentation team says about that. @kblaszczak-intel, @tsavina?

> Low-ranking Adaptation (LoRA) is a popular method to tune Generative AI models to a downstream task or custom data. But it requires some extra steps to be done for efficient deployment using the Hugging Face API. Namely, the trained adapters should be fused into the baseline model to avoid extra computation. This is how it can be done for LLMs:

Reviewer 5: We can add more links at the end of the article, in "Additional Resources." This way we don't distract readers when they should focus on our content but we give them the reference when they are done.

.. code-block:: python

    model_id = "meta-llama/Llama-2-7b-chat-hf"
    lora_adaptor = "./lora_adaptor"

    model = AutoModelForCausalLM.from_pretrained(model_id, use_cache=True)
    model = PeftModelForCausalLM.from_pretrained(model, lora_adaptor)
    model.merge_and_unload()
    model.get_base_model().save_pretrained("fused_lora_model")

> Now the model can be converted to OpenVINO using Optimum-Intel Python API or CLI interfaces mentioned above.

Reviewer 4: ```suggestion
Now the model can be converted to OpenVINO using Optimum Intel Python API or CLI interfaces mentioned above.
```

> Running Generative AI models using native OpenVINO APIs 

Reviewer 4: ```suggestion
Running Generative AI Models using Native OpenVINO APIs 
```

########################################################

> To run Generative AI models using native OpenVINO APIs you need to follow regular **Сonvert -> Optimize -> Deploy** path with few simplifications. 

Reviewer 4: ```suggestion
To run Generative AI models using native OpenVINO APIs, follow the regular **Сonvert -> Optimize -> Deploy** path with few simplifications. 
```

> To convert model from Hugging Face you can use Optimum-Intel export feature that allows to export model in OpenVINO format without invoking conversion API and tools directly, as it is shown above. In this case, the conversion process is a bit more simplified. You can still use regular conversion path if model comes from outside of Hugging Face ecosystem, i.e., in source framework format (PyTorch, etc.) 

Reviewer 4: ```suggestion
To convert a model from Hugging Face, you can use the Optimum Intel export feature, which enables model export in the OpenVINO format without directly invoking conversion API and tools, as shown above. This simplifies the conversion process. You can still use the regular conversion path if the model comes from outside the Hugging Face ecosystem, such as in a source framework format (PyTorch, etc.). 
```

> Model optimization could be performed within Hugging Face or directly using NNCF as described :doc:`here <weight_compression>`.

Reviewer 4: ```suggestion
Model optimization can be performed within Hugging Face or directly using NNCF as described in the :doc:`weight compression guide <weight_compression>`.
```

Reviewer: Link to HF docs for model optimization: https://huggingface.co/docs/optimum/intel/optimization_ov (weights compression will be added soon)

> Inference code that uses native API cannot benefit from Hugging Face pipelines, hence you need to write your own or take it from the examples available. Below are few examples of popular Generative AI scenarios:

Reviewer 4: ```suggestion
Inference code that uses native API cannot benefit from Hugging Face pipelines. You need to write your custom code or take it from the available examples. Below are some examples of popular Generative AI scenarios:
```

* In case of text Large Language Models (LLMs), this includes tokenization, inference and token selection loop, and de-tokenization. In case if token selection involves beam search, this also needs to be written.  

> * For Image generation models you need make a pipeline that includes several model: inference for source (e.g. text) encoder models, inference loop for diffusion process and inference for decoding part. Scheduler code is also required. 

Reviewer 4: ```suggestion
* For image generation models, you need to make a pipeline that includes several model inferences: inference for source (e.g., text) encoder models, inference loop for diffusion process and inference for decoding part. Scheduler code is also required. 
```

Reviewer 4: not sure what is meant by (e.g., text) here

> To write such pipelines, we recommend following examples that we have provided as a part of OpenVINO: 

Reviewer 4: ```suggestion
To write such pipelines, you can follow the examples provided as part of OpenVINO: 
```
